File: python/graph_general/course_schedule.py
from typing import List
from collections import deque

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = {i: [] for i in range(numCourses)}

        in_degree = [0] * numCourses

        for course, preq in prerequisites:
            graph[preq].append(course)

            in_degree[course] += 1


        queue = deque([i for i in range(numCourses) if in_degree[i] == 0])
        completed_courses = 0

        while queue:
            course = queue.popleft()

            completed_courses += 1

            for next_course in graph[course]:
                in_degree[next_course] -= 1

                if in_degree[next_course] == 0:
                    queue.append(next_course)


        return completed_courses == numCourses

File: golang/graph_general/course_schedule.go
package main

func canFinish(numCourses int, prerequisites [][]int) bool {
	graph := make(map[int][]int)
	inDegree := make([]int, numCourses)

	for _, pre := range prerequisites {
		course, prereq := pre[0], pre[1]
		graph[prereq] = append(graph[prereq], course)
		inDegree[course]++
	}

	queue := []int{}

	for i := 0; i < numCourses; i++ {
		if inDegree[i] == 0 {
			queue = append(queue, i)
		}
	}

	completedCourses := 0

	for len(queue) > 0 {
		course := queue[0]
		queue = queue[1:]
		completedCourses++

		for _, nextCourse := range graph[course] {
			inDegree[nextCourse]--

			if inDegree[nextCourse] == 0 {
				queue = append(queue, nextCourse)
			}
		}
	}

	return completedCourses == numCourses

}
